source/Calamari/Kubernetes/KubernetesContextScriptWrapper.cs

```diff
@@ -1,79 +1,717 @@
+                               LogType.Info,
+                               "cloud",
+                               "set",
+                               "--name",
+                               GetAzEnvironment());
+
+                log.Verbose("Azure CLI: Authenticating with Service Principal");
+
+                var subscriptionId = variables.Get("Octopus.Action.Azure.SubscriptionId");
+                var tenantId = variables.Get("Octopus.Action.Azure.TenantId");
+                var clientId = variables.Get("Octopus.Action.Azure.ClientId");
+                var password = variables.Get("Octopus.Action.Azure.Password");
+
+                ExecuteCommand(az,
+                               LogType.Info,
+                               "login",
+                               "--service-principal",
+                               // Use the full argument with an '=' because of https://github.com/Azure/azure-cli/issues/12105
+                               $"--username=\"{clientId}\"",
+                               $"--password=\"{password}\"",
+                               $"--tenant=\"{tenantId}\"");
+
+                log.Verbose($"Azure CLI: Setting active subscription to {subscriptionId}");
+                ExecuteCommand(az,
+                               LogType.Info,
+                               "account",
+                               "set",
+                               "--subscription",
+                               subscriptionId);
+
+                log.Info("Successfully authenticated with the Azure CLI");
+            }
+
+            string ConfigureCliExecution()
+            {
+                var kubeConfig = Path.Combine(workingDirectory, "kubectl-octo.yml");
+
+                // create an empty file, to suppress kubectl errors about the file missing
+                File.WriteAllText(kubeConfig, string.Empty);
+
+                environmentVars.Add("KUBECONFIG", kubeConfig);
+
+                if (scriptSyntax == ScriptSyntax.Bash)
+                {
+                    ExecuteCommand("chmod", LogType.Verbose, "u=rw,g=,o=", kubeConfig);
+                }
+
+                log.Verbose($"Temporary kubectl config set to {kubeConfig}");
+
+                return kubeConfig;
+            }
+
+            bool TrySetKubectl()
+            {
+                kubectl = variables.Get("Octopus.Action.Kubernetes.CustomKubectlExecutable");
+                if (string.IsNullOrEmpty(kubectl))
+                {
+                    kubectl = CalamariEnvironment.IsRunningOnWindows
+                        ? ExecuteCommandAndReturnOutput("where", "kubectl.exe").FirstOrDefault()
+                        : ExecuteCommandAndReturnOutput("which", "kubectl").FirstOrDefault();
+
+                    if (string.IsNullOrEmpty(kubectl))
+                    {
+                        log.Error("Could not find kubectl. Make sure kubectl is on the PATH. See https://g.octopushq.com/KubernetesTarget for more information.");
```

Comment:
See the comment above about linking to docs on error.

This error in particular was reported as an issue for a customer: https://octopusdeploy.slack.com/archives/C0GKZRNP4/p1621848449006000